### DISCUSSION TITLE
Interfaces to VPSC  or other crystal plasticity codes

### DISCUSSION THREAD

**Discussion by ralfHielscher** [mtex-toolbox/mtex#886] [Feature Requests]
Allow to call VPSC and other crystal plasticity codes from within MTEX

first model: we should distinguish mandatory / optional parameters

class **plasticityModel**
contains the plasticity / elasticity parameters for a single phase
- [ ] crystal symmetry
- [ ] slipSystems with hardening coefficients
- [ ] twinSystems
- [ ] elastic tensors
- [ ] thermal expansion

class **deformationModel**
describes the imposed deformation path 
- [ ] velocity gradient
- [ ] time steps
- [ ] temperature
- [ ] boundary conditions
- [ ] external stress
- [ ] some check?

class **microstructure**
describes the state of the microstructure at a certain time
- [ ] orienations / ODF describing the texure
- [ ] grains / fitted ellipses describing the mircrostructure, 2d -> 3d reconstruction
- [ ] internal stress
- [ ] temperature distribution 
- [ ] disclocation density

class **plasticSimulation**
represents the entire simulation, i.e, parameters, initial texture, result
- [ ] list of plasticityModels one for each phase
- [ ] deformationModel -> times
- [ ] simulation method: Taylor, affine, secant, second order, viscoplastic
- [ ] list of microstructures, one for each time step
 
The procedure would not be
1. set up a variable of class **plasticSimulation** with
* the plasticity models of each phase
* the deformation we want to simulate
* the initial microstructure
* the time steps
2. call VPSC
3. the result will be written directly into the class, it will then contain a list of microstructures one for each time step

**Comment by lfgmorales** [comment]
Hi Ralf

I can try to help with this one, that would be really cool

Cheers

Luiz

**Comment by ralfHielscher** [comment]
Some first work on importing results from VPSC has been done in f5644bc700f2ff2e86d25b9562cb83149339b904
and are illustrated in https://mtex-toolbox.github.io/VPSCImport.html

https://github.com/tori-miller/Ti64-VPSC-CoRotation-Analysis looks very promising. I already contacted Victoria Miller
for future cooperation.

**Reply by kilir** [reply]
00d0d05b8b617cc81b5116f71a76bd470611a95a : 
- adding the individual orientations allows to compute odfs with a individual kernels.
- adding the weights (although I've never seen them differ)
- the example file Luiz send has additional data, according to the VPSC manual there might be up to 4 addtional columns
 (vM strain, vM stress, total plastic work, Taylor factor), however those can be switched on or off individually

**Comment by tori-miller** [comment]
We (@Blackbelt1221) and I have done a lot of brainstorming because we had planned to make a VPSC toolbox independently of MTEX; however, incorporating it into MTEX will be much easier! These were our original plans, but we'd love to hear some discussion or better ways to do this. (Also this was originally typed for a slightly different audience!)

VPSC is an executable. Notably the creators of VPSC have not posted it publicly and request that you contact them if you want a copy, so I would recommend that we make this package “bring your own VPSC executable.” Alternatively, we could certainly ask Lebensohn and Tome whether we could include an executable in this package.

The VPSC executable requires a number of text files to run. At minimum, this will be: an infile named vpsc7.in, texture files for each phase, single crystal parameter files for each phase, and deformation operation files for each operation. There may also be additional files, such as grain morphology files for each phase. Each of these files has a standardized and fairly rigid format, though there are parts of the VPSC source code that may be commented or uncommented for some users which slightly changes the outputs.

As vpsc is simply an executable, you simply call it. You can either call it using system, dos, or unix as appropriate. You must place and call the executable in the directory where the input files are located and where the output files will be generated. You can also specify relative file paths for the locations of the (other) input files within the infile, but vpsc7.in must be located in the same place as the executable. For our convenience, we have preferred to standardize our file extension usage for all of the input files; however, they are arbitrary.

VPSC also outputs a number of text files, specifically: stepwise relative slip system activity (for each phase), stepwise stress-strain (tensorial and Von Mises), texture as a list of grain orientations for each phase. Depending on settings, the texture files may contain outputs at multiple steps within the deformation process. There are also a number of log files RERR.OUT, RUNLOG.OUT, STATS.OUT. In our experience, these are not typically needed by the user. If you specified grain morphologies as an input, you will also get final grain morphologies as an output (ellipsoid aspect ratios and euler angles).

VPSC has a number of additional, more specialized capabilities (polycrystalline yield surfaces, components of cubic textures, Lankford coefficients) that we would suggest are a secondary priority for implementation.  Similarly, there are different constitutive models (MTS) that require different parameters. However, I have not seen this widely used and I would also think of it as a lower priority. 

There is also a capability called POSTMORT which preserves the output of one VPSC run to be used as the input of a second VPSC run. In our experience, it was far better and more flexible to set this up programmatically using MTEX rather than relying on VPSC.

Running VPSC using MTEX requires an unfortunate amount of reading from and writing to text files. Luckily, we have nearly all of these generally framed out, at least for the parameters that we have needed to update. We will simply need to make very generalized versions.
 
**Input files:** 
-vpsc7.in: This contains both text and numerical data, and there are not a fixed number of inputs because you can run an arbitrary number of deformation steps. We definitely think this should be a new class. 
-single crystal files: these files are very information dense. They include: crystal symmetry, lattice parameters, elastic stiffness (not actually needed but is included), thermal expansion coefficients (unused), slip and twinning modes + flags for which are used for this simulation, lists of planes and directions for each slip/twinning mode, flags for constitutive law and if rate sensitive, grain size, and Voce/latent hardening parameters for each enabled slip or twinning mode. There are additional parameters for twin reorientation thresholds (how much twinning is needed before orientation of grain should be altered). These require parameters from the MTEX crystal symmetry class, as well as slipSystem and twinningSystem. Additionally, it will require storing many scalar parameters for each slip or twinning system. We could consider giving slip and twinning systems in MTEX optional parameters for VPSC-compatible hardening. We also think that it makes sense to have a new “SX file” class to hold all of this information. 
-deformation files: multiple types. One type is simply a rotation matrix, very simple. We already have a basic implementation of this for VPSC. The other main type is a velocity gradient. This consists of 4 scalars (number of steps, stepwise strain, temperature, and direction of strongest imposed component), 2 logical matrices (imposed or calculated), and 2 numerical matrices (velocity gradient and Cauchy stress). These files could also be made a class so that it is easy to implement a function to write the file. 

**Input and output:** 
-Texture: list of orientations and weights, also contains the number of grains and a flag for which Euler angle convention is in use. This would be easy to store as a list of MTEX orientations, plus weights for each grain. Also should contain the value “texture at strain =”, a flag for pre or post VPSC run, number of grains, and Euler angle convention.  We planned to make this a (trivial) class for consistency in file handling. 
-Morphological files: rotations types and axes lengths. 

**Output files:** 
-Slip activity: Fractional activity for each possible slip/twin mode in each column, plus a column per strain and a column for the average number of slip systems active all grains. Probably just want to easily be able to plot this as a function of strain. 
-Stress strain:  Tensorial components and Von Mises at each strain step. Probably want to be able to read this in and easily plot the VM str-str and represent tensorial components as an MTEX tensor. 

We have also discussed setting up an overall VPSC class that could hold both the inputs and outputs, and include the functions to run VPSC and read the outputs within it, but this has the potential to result in users overwriting the input properties but not deleting/updating the outputs. Implementing a warning would be critical. 

As you can see, our plan requires very little alteration to the core MTEX functionality--just giving slipSystems the ability to store a few extra optional parameters. Alternatively we could set up vpscSlipSystem as a subclass and not touch core MTEX. 

This is how we (@Blackbelt1221) had planned to build the toolbox when we were going to develop it independently of MTEX. Of course, we would love to draw for your experience about better ways to structure this. We thought we’d put our original plans out here as a starting point for discussion.

**Reply by ralfHielscher** [reply]
Hi Tori,

thank you very much for this extensive description. I updated the first post to represent a first 
idea how to represent the values you mention. Every comment is welcome.

Ralf

**Reply by tori-miller** [reply]
Hi Ralf, 
What you have above looks excellent! A few initial comments based on what you've written: 

- the vpsc code actually contains interaction models beyond just vpsc. It can additionally do Taylor, affine, secant, and second order interactions. This is changed using a single parameter in the infile, so it could be easily included in plasticSimulation. 
- the deformationModel class as you currently have it described only allows for strain-controlled deformation. Some users may want to do stress-controlled deformation or a simulation with mixed boundary conditions. To enable that, we'd need to include a second tensor for stress and some way to flag which boundary conditions are fixed, e.g. a logic matrix. 
- I have some (minor) hesitation about calling the class just microstructure, as to me that also implies positional information. For new users of VPSC, one of the most common errors I see is the assumption that VPSC contains spatial info, and I am afraid that calling the class microstructure may further that confusion. That being said, I'm struggling to come up with a compact alternative. 
- users will certainly want to take their ebsd microstructures and make a representative vpsc microstructure for plastic simulation.  One of the challenges in generalizing and automating this is in converting from 2d ellipse aspect ratios to 3d ellipsoid aspect ratios. We will need to be careful about what assumptions we make here when filling in the 3rd dimension. 

Thanks for all your work on this! 
Tori.

**Reply by Begley-Dev** [reply]
Hi Ralf,

Everything looks good! I have one additional comment that is not on Tori's list. When using VPSC for a strain-controlled test rather than a stress-controlled test, VPSC increments over strain steps, rather than time steps. We may want to implement some sort of check in the deformationModel class that can reduce some of that confusion, because VPSC will run regardless of whether you pass it the right kind of increment.

Ben

**Reply by ralfHielscher** [reply]
thank you for your input. I tried to update the post accordingly. 
I also created a new branch https://github.com/mtex-toolbox/mtex/tree/feature/VPSCInterface where the development should take place and made started setting up the new classes.

I think next step should be to have a minimal working example, e.g.,
only slipSystems, only Taylor model, only orientations, 

Ralf.
